### PR TITLE
feat(aws-lambda) add S3BatchEvent

### PR DESCRIPTION
**Pull request by pwmcintyre**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.aws.amazon.com/AmazonS3/latest/dev/batch-ops-invoke-lambda.html
- [ ] (Not applicable) If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] (Not applicable) If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by pwmcintyre** [comment]
seeking feedback / review from recent contributors: @juancarbonel @redlickigrzegorz @skyzenr @loikg @dalen

**Review by skyzenr — APPROVED** [review]
Looks good to me!

**Comment by amcasey** [comment]
@andrewbranch, this looks good to me but the completion perf number is superficially concerning.  What is that testing?  Would it be sensitive to the fact that there are now more symbols sharing a prefix?

File: types/aws-lambda/index.d.ts [hunk 1]
```diff
@@ -252,6 +253,46 @@ export interface S3Event {
...
+}
+
+export interface S3BatchEventJob {
+    id: string;
+}
+
+export interface S3BatchEventTask {
+    taskId: string;
+    s3Key: string;
+    s3VersionId?: string;
```
Comment (simonbuchan) [inline types/aws-lambda/index.d.ts]: This being optional makes me a little curious, as it isn't mentioned in the docs: is it completely not present for non-versioned buckets?

**Review by simonbuchan — CHANGES_REQUESTED** [review]
This looks pretty great, but it's missing the specialized handler and context type, down the bottom of `index.d.ts`. I should probably clean this up at some point to make things like that more obvious....

File: types/aws-lambda/index.d.ts (hunk 1 above)
Comment (pwmcintyre) [reply types/aws-lambda/index.d.ts]: yeah that's a really good point -- all my buckets are actually version enabled, let me whip something up

Comment (pwmcintyre) [reply types/aws-lambda/index.d.ts]: @simonbuchan 

okay the results are in -- and you're correct, it's `null` -- i've committed a fix

actual payload:
```
            {
                "taskId": "AAAAAAAAAAHLSJmnbBR1fLJ5Y9/truncated",
                "s3BucketArn": "arn:aws:s3:::ommitted",
                "s3Key": "47600731-b366-4444-b0f4-24255c08c6e3.json",
                "s3VersionId": null
            }
```

**Comment by pwmcintyre** [comment]
@simonbuchan -- addressed, see latest commit

> This looks pretty great, but it's missing the specialized handler and context type, down the bottom of `index.d.ts`. I should probably clean this up at some point to make things like that more obvious....

File: types/aws-lambda/index.d.ts (hunk 1 above)
Comment (simonbuchan) [reply types/aws-lambda/index.d.ts]: Thanks for checking!

**Comment by simonbuchan** [comment]
Thanks for the investigation and fixes!

Sorry for the pedentry, but I also just noticed looking at the handler types you're using `...EventRequest` and `...EventResponse`, the convention is that the actual handler input and result types are `...Event` and `...Result`, so for this `S3BatchEvent` and `S3BatchResult` (and similar updates for the child types).

There are probably some `...EventRequest` and `...EventResponse` types in there for `request` and `response` properties of `...Event`, so I'd like to keep these consistent to avoid confusion. It does unfortunately suggest the `results` property of `S3BatchResult` should have type `S3BatchResultResult[]`, though 🤷‍♂️

**Comment by pwmcintyre** [comment]
@simonbuchan -- totally fair points, please review latest commits. It does get a little quirky with the `S3BatchResultResultCode` ... but it does hold true to the convention, so I think that's how it should stay.

> Thanks for the investigation and fixes!
> 
> Sorry for the pedentry, but I also just noticed looking at the handler types you're using `...EventRequest` and `...EventResponse`, the convention is that the actual handler input and result types are `...Event` and `...Result`, so for this `S3BatchEvent` and `S3BatchResult` (and similar updates for the child types).
> 
> There are probably some `...EventRequest` and `...EventResponse` types in there for `request` and `response` properties of `...Event`, so I'd like to keep these consistent to avoid confusion. It does unfortunately suggest the `results` property of `S3BatchResult` should have type `S3BatchResultResult[]`, though 🤷‍♂️

**Review by simonbuchan — APPROVED** [review]
Looks great! Thanks again for the quick turnaround.

**Pull request merged**